.ci/Treecheck/Treecheck.jl

```diff
@@ -0,0 +1,187 @@
+    end
+
+    # Necessary to avoid erroring on empty matrix:
+    if isempty(dict["include"])
+        json_str = ""
+    else
+        json_str = JSON3.write(dict)
+    end
+
+    println(Base.stderr, json_str)
+
+    gha_set_output("matrix", json_str)
+
+    return json_str
+end
+
+
+
+function _treehash_to_commithash(treehash::AbstractString)
+    cmd_1 = `git log --pretty=raw`
+    cmd_2 = `grep -B 1 "$(treehash)"`
+    my_pipeline = pipeline(cmd_1, cmd_2)
+    str = strip(_tryread(my_pipeline))
+    r = r"commit ([A-Za-z0-9]*?)\n"
+    m = match(r, str)
+    commithash = strip(m[1])
+    return commithash
+end
+
+function _tryread(cmd::Base.AbstractCmd)
+    str = try
+        read(cmd, String)
+    catch ex
+        # If an exception is thrown, run the command again, but send stdout/stderr to the CI log
+        run(cmd)
+
+        rethrow()
+    end
+    return str
+end
+
+check(package_uuid_str::AbstractString) = check(RegistryToml(), package_uuid_str)
+function check(registrytoml::RegistryToml, package_uuid_str::AbstractString)
+    package_name = registrytoml.dict["packages"][package_uuid_str]["name"]
+    package_relpath = registrytoml.dict["packages"][package_uuid_str]["path"]
+    package_abspath = joinpath(registrytoml.root, package_relpath)
+    package_toml = joinpath(package_abspath, "Package.toml")
+    versions_toml = joinpath(package_abspath, "Versions.toml")
+    package_dict = TOML.parsefile(package_toml)
+    versions_dict = TOML.parsefile(versions_toml)
+    package_git_repo_url = package_dict["repo"]
+    @testset "Treecheck for package: $(package_name)" begin
+        @test !isempty(versions_dict)
+        mktempdir() do tmpdir
+            cd(tmpdir) do
+                run(`git clone "$(package_git_repo_url)" MYCLONEDIR`)
+                cd("MYCLONEDIR") do
+                    gitrepo_libgit2 = LibGit2.GitRepo(".")
+
+                    # For each tree, make sure the tree exists
+                    @testset for (k, v) in pairs(versions_dict)
+                        treehash = v["git-tree-sha1"]
+                        tree_libgit2 = LibGit2.GitTree(gitrepo_libgit2, LibGit2.GitHash(treehash))
+                        @test tree_libgit2 isa LibGit2.GitTree
+                    end
+
+                    treehash_to_commithash = Dict{String, String}()
+
+                    # For each tree, make sure that at least one commit points to that tree
+                    @testset for (k, v) in pairs(versions_dict)
+                        treehash = v["git-tree-sha1"]
+                        commithash = _treehash_to_commithash(treehash)
+                        @test !isempty(strip(commithash))
+                        r = r"^[A-Za-z0-9]*?$"
+                        @test occursin(r, commithash)
+
+                        treehash_to_commithash[treehash] = commithash
+                    end
+
+                    # For each commit:
+                    # 1. Make sure that `git checkout COMMIT` does not error
```

Comment:
In which scenario would it error? Why does it matter?

Reply:
Moot after https://github.com/JuliaRegistries/General/commit/3209879f8014c2d606ec3dfedb57b4df67b77673, I think